src/detectors.rs

```diff
@@ -0,0 +1,340 @@
+
+const PANIC_WORKER_BLOCK_DURATION_DEFAULT: Duration = Duration::from_secs(60);
+
+fn get_panic_worker_block_duration() -> Duration {
+    let duration_str = env::var("MY_DURATION_ENV").unwrap_or_else(|_| "60".to_string());
+    duration_str
+        .parse::<u64>()
+        .map(Duration::from_secs)
+        .unwrap_or(PANIC_WORKER_BLOCK_DURATION_DEFAULT)
+}
+
+#[cfg(unix)]
+fn get_thread_id() -> libc::pthread_t {
+    unsafe { libc::pthread_self() }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct ThreadInfo {
+    id: ThreadId,
+    #[cfg(unix)]
+    pthread_id: libc::pthread_t,
+}
+
+impl ThreadInfo {
+    fn new() -> Self {
+        ThreadInfo {
+            id: thread::current().id(),
+            #[cfg(unix)]
+            pthread_id: get_thread_id(),
+        }
+    }
+
+    // Getter for the id field
+    pub fn id(&self) -> &ThreadId {
+        &self.id
+    }
+
+    // Getter for the pthread_id field (only available on Unix)
+    #[cfg(unix)]
+    pub fn pthread_id(&self) -> &libc::pthread_t {
+        &self.pthread_id
+    }
+}
+
+/// A trait for handling actions when blocking is detected.
+///
+/// This trait provides a method for handling the detection of a blocking action.
+pub trait BlockingActionHandler: Send + Sync {
+    /// Called when a blocking action is detected and prior to thread signaling.
+    ///
+    /// # Arguments
+    ///
+    /// * `workers` - The list of thread IDs of the tokio runtime worker threads.   /// # Returns
+    ///
+    fn blocking_detected(&self, workers: &[ThreadInfo]);
+}
+
+impl<F> BlockingActionHandler for F
+where
+    F: Fn(&[ThreadInfo]) + Send + Sync,
+{
+    fn blocking_detected(&self, workers: &[ThreadInfo]) {
+        // Implement the behavior for blocking_detected using the provided function.
+        // You can call the function here with the given workers.
+        self(workers);
+    }
+}
+struct StdErrBlockingActionHandler;
+
+/// BlockingActionHandler implementation that writes blocker details to standard error.
+impl BlockingActionHandler for StdErrBlockingActionHandler {
+    fn blocking_detected(&self, workers: &[ThreadInfo]) {
+        eprintln!("Detected blocking in worker threads: {:?}", workers);
+    }
+}
+
+#[derive(Debug)]
+struct WorkerSet {
+    inner: Mutex<HashSet<ThreadInfo>>,
+}
+
+impl WorkerSet {
+    fn new() -> Self {
+        WorkerSet {
+            inner: Mutex::new(HashSet::new()),
+        }
+    }
+
+    fn add(&self, pid: ThreadInfo) {
+        let mut set = self.inner.lock().unwrap();
+        set.insert(pid);
+    }
+
+    fn remove(&self, pid: ThreadInfo) {
+        let mut set = self.inner.lock().unwrap();
+        set.remove(&pid);
+    }
+
+    fn get_all(&self) -> Vec<ThreadInfo> {
+        let set = self.inner.lock().unwrap();
+        set.iter().cloned().collect()
+    }
+}
+
+/// Utility to help with "really nice to add a warning for tasks that might be blocking"
+#[derive(Debug)]
+pub struct LongRunningTaskDetector {
```

Comment:
This documentation isn't great. Can't you explain what it does instead? Same for the module-documentation. (It's okay if you copy the same documentation here and on the module.)